web-api/src/storage/elastic.rs

```diff
@@ -145,72 +150,68 @@ impl Client {
     }
 
-    pub(super) async fn insert_documents(
+    pub(super) async fn upsert_documents(
         &self,
-        documents: impl IntoIterator<
-            IntoIter = impl ExactSizeIterator<Item = &models::DocumentForIngestion>,
-        >,
+        documents: &[models::DocumentForIngestion],
     ) -> Result<Warning<DocumentId>, Error> {
-        let documents = documents.into_iter();
-        if documents.len() == 0 {
-            return Ok(Warning::default());
-        }
-
-        let response = self
-            .bulk_request(documents.flat_map(|document| {
-                [
-                    serde_json::to_value(BulkInstruction::Index { id: &document.id })
-                        .map_err(Into::into),
-                    serde_json::to_value(IngestedDocument {
-                        // TODO[pmk/ET-4756-7]
-                        snippet: &document.snippets[0].0,
-                        properties: &document.properties,
-                        // TODO[pmk/ET-4756-7]
-                        embedding: &document.snippets[0].1,
-                        tags: &document.tags,
-                    })
-                    .map_err(Into::into),
-                ]
-            }))
-            .await?;
-        Ok(response.failed_documents("index", false).into())
+        let ids = documents.iter().map(|document| &document.id).collect_vec();
+        self.delete_by_parents(ids).await?;
+        self.freshly_insert_documents(documents).await
     }
 
-    pub(super) async fn delete_documents(
+    pub(super) async fn freshly_insert_documents(
         &self,
-        ids: impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = &DocumentId>>,
+        documents: impl IntoIterator<Item = &models::DocumentForIngestion>,
     ) -> Result<Warning<DocumentId>, Error> {
-        let ids = ids.into_iter();
-        if ids.len() == 0 {
+        // TODO[pmk/now] this won't work correctly with upserts id old snippet count > new snippet count
+        let mut snippets = documents
+            .into_iter()
+            .flat_map(|document| {
+                document
+                    .snippets
+                    .iter()
+                    .enumerate()
+                    .flat_map(|(idx, (snippet, embedding))| {
+                        #[allow(clippy::cast_possible_truncation)]
+                        let id = SnippetId::new(document.id.clone(), idx as _);
+                        let header =
+                            serde_json::to_value(BulkInstruction::Create { id: id.to_es_id() })
+                                .map_err(Into::into);
+                        let data = serde_json::to_value(EsDocument {
+                            snippet,
+                            properties: &document.properties,
+                            embedding,
+                            tags: &document.tags,
+                            parent: id.document_id(),
+                        })
+                        .map_err(Into::into);
```

Comment:
:art: no need for extra error conversion
```suggestion
                        });
```